### PR TITLE
Allow custom `jaxley` solvers to have different nseg per branch

### LINKED ISSUE #429: The change in capacitance takes effect only at the branch(0), when len(parents) >1.
The following codes will show the same result when giving a step current to branch(0):
A:
branch = jx.Branch(comp, nseg=1)
cell = jx.Cell(branch, parents=[-1,0])
cell.branch(1).set('capacitance',1)
B:
branch = jx.Branch(comp, nseg=1)
cell = jx.Cell(branch, parents=[-1,0])
cell.branch(1).set('capacitance',200)

### PR DESCRIPTION
**Pull request by michaeldeistler**
Following #418, this will also allow different nseg per branch for the jaxley.thomas solver. This works by masking the tridiagonal system in every branch.

In addition, it also uses more invasively `.comp_edges` to build the conductances of the custom solver more easily. Before this PR, `.comp_edges` had only been used for the `jax.sparse` solver.

Replaces #425 which was closed due to difficult merge conflicts.

In addition, this contains a bugfix and closes #429.

### CODE REVIEW

File: jaxley/solver_voltage.py [hunk 1]
```diff
@@ -89,24 +88,60 @@ def step_voltage_implicit_with_custom_spsolve(
...
+        )
+
+    # Build lowers.
+    lowers = jnp.zeros(nbranches * nseg)
+    lower_inds = sources[c2c] < sinks[c2c]
+    sinks_lower = sinks[c2c][lower_inds]
+    if len(sinks_lower) > 0:
+        lowers = lowers.at[masked_node_inds[sinks_lower]].add(
+            -delta_t * axial_conductances[c2c][lower_inds]
+        )
```
[the remark this answers is not in the document]
Comment (jnsbck) [reply jaxley/solver_voltage.py:129]: saves a few lines, but not necessarily more readble

File: jaxley/solver_voltage.py [hunk 2]
```diff
@@ -213,24 +250,22 @@ def step_voltage_implicit_with_jax_spsolve(
...
 ):
+    axial_conductances = delta_t * axial_conductances
+
     # Build diagonals.
     diagonal_values = jnp.zeros(n_nodes)
 
     # if-case needed because `.at` does not allow empty inputs, but the input is
     # empty for compartments.
     if len(sinks) > 0:
-        diagonal_values = diagonal_values.at[sinks].add(delta_t * axial_conductances)
```
Comment (jnsbck) [inline jaxley/solver_voltage.py:222]: This is different from L99

File: jaxley/utils/cell_utils.py [hunk 3]
```diff
@@ -409,95 +409,29 @@ def query_channel_states_and_params(d, keys, idcs):
...
+            params["length"][sink_comp_inds],
+            params["length"][source_comp_inds],
+        )
+    else:
+        conds_c2c = jnp.asarray([])
 
-    # `branchpoint-to-compartment` (bp2c) conductances.
-    condition = comp_edges["type"].to_numpy() == 1
+    # `branchpoint-to-compartment` (bp2c) axial coupling conductances.
+    condition = np.isin(comp_edges["type"].to_numpy(), [1, 2])
```
Comment (jnsbck) [inline jaxley/utils/cell_utils.py:441]: again could be done with `df.isin`

File: jaxley/utils/solver_utils.py [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+
+
+def remap_index_to_masked(
+    index, nodes: pd.DataFrame, max_nseg: int, nseg_per_branch: jnp.ndarray
+):
+    """Convert actual index of the compartment to the index in the masked system.
+
+    E.g. if `nsegs = [2, 4]`, then the index `3` would be mapped to `5` because the
+    masked `nsegs` are `[4, 4]`.
+    """
```
Comment (jnsbck) [inline jaxley/utils/solver_utils.py:18]: Took me a sec to understand what you meant. Maybe you can add this as well 
```
orig:   [0, 1,           2, 3, 4, 5] 
masked: [0, 1, (2) ,(3) ,4, 5, 6, 7]
```.

**Review by jnsbck — APPROVED** [review]
Nice refactor. I like the line diff! :dagger: Left a few comments. If tests pass and there are no issues, good to go from my side.

File: jaxley/solver_voltage.py (hunk 2 above)
Comment (michaeldeistler) [reply jaxley/solver_voltage.py:222]: It's all good. L99 initializes with `diags = jnp.ones(nbranches * nseg)` (which corresponds to `1+...`) and here we premultiply with `delta_t` in line 253.

File: jaxley/modules/compartment.py [hunk 5]
```diff
@@ -89,18 +92,10 @@ def _init_morph_jax_spsolve(self):
...
-            "branchpoint_conds_parents": jnp.asarray([]),
-            "branchpoint_weights_children": jnp.asarray([]),
-            "branchpoint_weights_parents": jnp.asarray([]),
-            "branch_uppers": jnp.asarray([]),
-            "branch_lowers": jnp.asarray([]),
-            "branch_diags": jnp.asarray([0.0]),
-        }
-
-    def _init_conds_jax_spsolve(self, params):
+    def init_conds(self, params: Dict[str, jnp.ndarray]):
```
[the remark this answers is not in the document]
Comment (michaeldeistler) [reply jaxley/modules/compartment.py:95]: good catch, should have been `_compute_axial_conductances`

File: jaxley/modules/base.py [hunk 6]
```diff
@@ -35,6 +34,7 @@
 from jaxley.utils.debug_solver import compute_morphology_indices
```
[the remark this answers is not in the document]
Comment (michaeldeistler) [reply jaxley/modules/base.py:34]: It is not used, I will keep it in debug

File: jaxley/modules/cell.py [hunk 7]
```diff
@@ -157,11 +149,13 @@ def __getattr__(self, key: str):
...
+    def _init_morph_jaxley_spsolve(self):
         """Initialize morphology for the custom sparse solver.
 
         Running this function is only required for custom Jaxley solvers, i.e., for
-        `voltage_solver={'jaxley.stone', 'jaxley.thomas'}`.
+        `voltage_solver={'jaxley.stone', 'jaxley.thomas'}`. However, because at
+        `.__init__()` (when the function is run), we do not yet know which solver the
+        user will use. Therefore, we always run this function at `.__init__()`.
         """
         children_and_parents = compute_morphology_indices_in_levels(
```
[the remark this answers is not in the document]
Comment (michaeldeistler) [reply jaxley/modules/cell.py:160]: I am not yet fully sure if it is possible, but we can think about it in the future.

File: jaxley/modules/network.py [hunk 8]
```diff
@@ -155,19 +151,34 @@ def _init_morph_custom_spsolve(self):
...
+
     def _init_morph_jax_spsolve(self):
         """Initialize the morphology for networks.
 
-        The reason that this is a bit involved is that Jaxley considers branchpoint
-        nodes to be at the very end of __all__ nodes (i.e. the branchpoints of the
-        first cell are even after the compartments of the second cell. The reason for
-        this is that, otherwise, `cumsum_nseg` becomes tricky).
+        The reason that this function is a bit involved for a `Network` is that Jaxley
+        considers branchpoint nodes to be at the very end of __all__ nodes (i.e. the
```
[the remark this answers is not in the document]
Comment (michaeldeistler) [reply jaxley/modules/network.py:167]: I don't think it creates a lot of overhead. We can chat about it in the future.
